frame/assets/src/lib.rs

```diff
@@ -878,200 +980,350 @@ pub mod pallet {
+		/// instructions.
+		///
+		/// Emits `AssetStatusChanged` with the identity of the asset.
+		///
+		/// Weight: `O(1)`
+		#[pallet::weight(T::WeightInfo::force_asset_status())]
+		pub(super) fn force_asset_status(
+			origin: OriginFor<T>,
+			#[pallet::compact] id: T::AssetId,
+			owner: <T::Lookup as StaticLookup>::Source,
+			issuer: <T::Lookup as StaticLookup>::Source,
+			admin: <T::Lookup as StaticLookup>::Source,
+			freezer: <T::Lookup as StaticLookup>::Source,
+			#[pallet::compact] min_balance: T::Balance,
+			is_sufficient: bool,
+			is_frozen: bool,
+		) -> DispatchResultWithPostInfo {
+			T::ForceOrigin::ensure_origin(origin)?;
+
+			Asset::<T>::try_mutate(id, |maybe_asset| {
+				let mut asset = maybe_asset.take().ok_or(Error::<T>::Unknown)?;
+				asset.owner = T::Lookup::lookup(owner)?;
+				asset.issuer = T::Lookup::lookup(issuer)?;
+				asset.admin = T::Lookup::lookup(admin)?;
+				asset.freezer = T::Lookup::lookup(freezer)?;
+				asset.min_balance = min_balance;
+				asset.is_sufficient = is_sufficient;
+				asset.is_frozen = is_frozen;
+				*maybe_asset = Some(asset);
+
+				Self::deposit_event(Event::AssetStatusChanged(id));
+				Ok(().into())
+			})
+		}
+
+		/// Approve an amount of asset for transfer by a delegated third-party account.
+		///
+		/// Origin must be Signed.
+		///
+		/// Ensures that `ApprovalDeposit` worth of `Currency` is reserved from signing account
+		/// for the purpose of holding the approval. If some non-zero amount of assets is already
+		/// approved from signing account to `delegate`, then it is topped up or unreserved to
+		/// meet the right value.
+		///
+		/// NOTE: The signing account does not need to own `amount` of assets at the point of
+		/// making this call.
+		///
+		/// - `id`: The identifier of the asset.
+		/// - `delegate`: The account to delegate permission to transfer asset.
+		/// - `amount`: The amount of asset that may be transferred by `delegate`. If there is
+		/// already an approval in place, then this acts additively.
+		///
+		/// Emits `ApprovedTransfer` on success.
+		///
+		/// Weight: `O(1)`
+		#[pallet::weight(T::WeightInfo::approve_transfer())]
+		pub(super) fn approve_transfer(
+			origin: OriginFor<T>,
+			#[pallet::compact] id: T::AssetId,
+			delegate: <T::Lookup as StaticLookup>::Source,
+			#[pallet::compact] amount: T::Balance,
+		) -> DispatchResultWithPostInfo {
+			let owner = ensure_signed(origin)?;
+			let delegate = T::Lookup::lookup(delegate)?;
+
+			let key = ApprovalKey { owner, delegate };
+			Approvals::<T>::try_mutate(id, &key, |maybe_approved| -> DispatchResult {
+				let mut approved = maybe_approved.take().unwrap_or_default();
+				let deposit_required = T::ApprovalDeposit::get();
+				if approved.deposit < deposit_required {
+					T::Currency::reserve(&key.owner, deposit_required - approved.deposit)?;
+					approved.deposit = deposit_required;
+				}
+				approved.amount = approved.amount.saturating_add(amount);
```

Comment:
This is definitely not the standard API. The standard API is that any `amount` is what is the final set value when making this call.

Setting amount as zero would also be a replacement for a `cancel_approval` call.

Reply:
sure, but unfortunately the standard API is fundamentally flawed for front-running since altering the order to a notional `adjust_approved` and `transfer_approved` would result in different amounts being transferable. i can add `adjust_approved` of course, but you can easily achieve the same thing by batching `cancel_approved` with `approve_transfer`.